hpat/distributed_api.py

```diff
@@ -195,15 +196,14 @@ def gatherv_str_arr_impl(data):
 
             # displacements
             all_data = StringArray([''])  # dummy arrays on non-root PEs
-            displs = np.empty(1, np.int32)
-            displs_char = np.empty(1, np.int32)
+            displs = np.empty(n_loc, np.int32)
+            displs_char = np.empty(n_loc, np.int32)
 
```

Comment:
`displs` and `displs_char` are needed on root only and will be set properly in the following if-condition.
This is just a dummy for all other ranks to keep the code simple.

A proper solution would make `c_gatherv` accept `None` or `0` for non-root ranks. This might require some extra work like casting...

BTW: `n_loc` is definitely not the right value anyway.

Reply:
@fschlimb Yes, I left it accidentally, it's not needed at all. I can change it to 0 size array, since the size doesn't really matter, or we can even remove these two lines as redundant. I'm not sure I got how to update c_gatherv. It accepts void*, but in gatherv_overload we have to keep the type consistent, so if we assign something to displs it it should have the same type as in the root branch. Shall I remove these lines?

Reply:
I don't think you can remove them because you need something type-stable to pass it to `c_gatherv`. If a zero-sized array works that'd be great.

If numba does not automatically cast 0 to a pointer when the C-function expects a void* you can write an intrinsic which returns a `((void*)0)` and use that in the function you @jit/@overload. Maybe like this
```
@intrinsic
def mk_int32_null(typingctx):
    def codegen(context, builder, sig, args):
        zero = context.get_constant(types.int32, 0)
        return builder.inttoptr(zero, llvmlite.ir.IntType(32))
    return numba.types.int32, codegen
```

It's probably not worth the effort, though. You will probably need yet another layer between the caller and the ExternFunction. In particular if passing an empty array works we should just use that.

Reply:
It looks like else branch can be remove (I tried and test still passes). Inserting "print(my_arr)" into it shows empty array on non root ranks (i.e. it's not even needed to assign something in the else branch), looks like Numba default constructs the object, once it's able to infer the type? I failed to find where the magic comes from in the IR. I suggest keeping latest commit with zero size array which works too.